pkg/investigations/machinehealthcheckunterminatedshortcircuitsre/machinehealthcheckunterminatedshortcircuitsre.go

```diff
@@ -0,0 +1,487 @@
+func (i *Investigation) investigateDeletingMachine(ctx context.Context, machine machinev1beta1.Machine) error {
+	if machine.Status.NodeRef == nil {
+		notes := "machine stuck deleting with no node"
+		i.recommendations.addRecommendation(recommendationInvestigateMachine, machine.Name, notes)
+		return nil
+	}
+	node, err := i.getNodeForMachine(ctx, machine)
+	if err != nil {
+		notes := "machine's node could not be determined"
+		i.recommendations.addRecommendation(recommendationInvestigateMachine, machine.Name, notes)
+		return fmt.Errorf("failed to retrieve node for machine %q: %w", machine.Name, err)
+	}
+
+	stuck, duration := checkForStuckDrain(node)
+	if stuck {
+		notes := fmt.Sprintf("node %q found to be stuck draining for %s", node.Name, duration.Truncate(time.Second).String())
+		i.recommendations.addRecommendation(recommendationInvestigateNode, node.Name, notes)
+		return nil
+	}
+
+	notes := "unable to determine why machine is stuck deleting"
+	i.recommendations.addRecommendation(recommendationInvestigateMachine, machine.Name, notes)
+	return nil
+}
+
+// checkForStuckDrain makes a best-effort approximation at whether a node is stuck draining a specific pod
+func checkForStuckDrain(node corev1.Node) (bool, *time.Duration) {
+	if len(node.Spec.Taints) == 0 {
+		return false, nil
+	}
+
+	taint, found := findNoScheduleTaint(node)
+	if !found {
+		return false, nil
+	}
+
+	// TODO - Once CAD can access on-cluster metrics, we can query the `pods_preventing_node_drain` metric from prometheus
+	// to more accurately gauge if a node is truly stuck deleting, and what pod is causing it
+	drainDuration := time.Since(taint.TimeAdded.Time)
+	if drainDuration > 10*time.Minute {
+		return true, &drainDuration
+	}
+
+	return false, nil
+}
+
+// findNoScheduleTaint searches the node's taints to find one with effect: NoSchedule, if present.
+//
+// If none is present, an empty taint and 'false' are returned
+func findNoScheduleTaint(node corev1.Node) (corev1.Taint, bool) {
+	for _, taint := range node.Spec.Taints {
+		if taint.Effect == corev1.TaintEffectNoSchedule {
+			return taint, true
+		}
+	}
+	return corev1.Taint{}, false
+}
+
+// getMachineRole returns the role of the given machine, if present. If not found, an error is returned
+func (i *Investigation) getMachineRole(machine machinev1beta1.Machine) (string, error) {
+	role, found := machine.Labels[machineRoleLabel]
+	if !found {
+		return "", fmt.Errorf("expected label key %q not found", machineRoleLabel)
+	}
+	return role, nil
+}
+
+// getNodesForMachines retrieves the nodes for the given machines. Errors encountered are joined, but do not block the retrieval of other machines
+func (i *Investigation) getNodesForMachines(ctx context.Context, machines []machinev1beta1.Machine) ([]corev1.Node, error) {
+	nodes := []corev1.Node{}
+	var errs error
+	for _, machine := range machines {
+		node, err := i.getNodeForMachine(ctx, machine)
```

Comment:
If we just run 2 kubernetes API calls (even though it might retrieve 'too many nodes' first) this might be faster still instead of running the loop that gets nodes one-by-one maybe?
Just get all machines we care about - then get all nodes for all of those in a single call and filter on the result.

Reply:
Good call, that would save us (quite a few) API calls

Reply:
Updated this logic in the latest commit to retrieve all nodes first, then filter out the unneeded ones before returning